### ISSUE TITLE
EJB timer database high disk usage

### ISSUE THREAD

**Issue by plachance** [javaee/glassfish#21933]
## Environment Details
* GlassFish Version (and build number): GlassFish Server Open Source Edition  4.0  (build 89)
* JDK version: OpenJDK 1.7.0_121
* OS: Ubuntu 12.04
* Database: PostgreSQL 9.4.11

----------

## Problem Description
The EJB timer database in /opt/glassfish4/glassfish/domains/domain1/lib/databases/ejbtimer/log uses a lot of disk space (approx. 10 GB). There's ~9000 files of 1.1 MB each. I don't think it should use so much disk space.

## Steps to reproduce
Unknown. What could cause this behavior?

## Impact of Issue
I think it could be related to a very long start/stop of the domain (10 to 15 minutes).

**Comment by yaminikb** [comment]
@plachance Can you please try with latest version of GlassFish i.e 5.0?
http://download.oracle.com/glassfish/5.0/nightly/index.html

**Comment by plachance** [comment]
We are unable to find why the domain start/stop take long time. There's nothing special in the logs and we could'nt reproduce it in our development env. The app still works and we rarely need to restart the domain anyway. Having to load a 10 GB database at the domain start is the best hypothesis we have to date.

There's only 1 app deployed on this container. The app offer a RESTful service to send emails and act as a bounce mail handler.

Here's the timers we have in the app:
* BounceMailHandler: @Schedule(second = "0", minute = "*/5", hour = "*")
* Cleanup of outdated data: @Schedule(second = "0", minute = "*/1", hour = "*")
* More cleanup: @Schedule(second = "0", minute = "*/10", hour = "*")
* Reload of container properties to avoid restarting the application when we change a property: @Schedule(second = "0", minute = "*/1", hour = "*")
* Some reporting: @Schedule(hour = "0,4,8,10,12,14,16,20")
* Send emails in the queues (1 timer per queue, 2 queues total): @Schedule(second = "*", minute = "*", hour = "*")

I'm curious about what does the ejb timer database contain. We have a couple of timers that run each seconds on functions that can return anywhen between instantaneously and an hour later. Does each "skipped" timer are saved to this database to be rerun later?

**Comment by yaminikb** [comment]
@amykang2020 Can you please help evaluate this issue?

**Comment by amykang2020** [comment]
For EJB peresistent timer,  GlassFish server stores the timer information in the database.  The information stored are shown in the glassfish/lib/install/databases/ejbtimer_<db-type>.sql file.   As it shows,  among the stored information it has timer schedule, timer state, lastexpiration, etc.  From these information the EJB timer service determines the next timeout.  

One thing it's worth to check is to see if the application creates persistent timers on-fly ..

**Comment by plachance** [comment]
Thanks for your suggestions. I found the problem and a solution. Here's what I found.

The application doesn't create programmatic timers. I don't think it creates EJBs annotated with **@schedule** on the fly. Most annotated EJBs are **@singleton**. There's only one **@Stateless** EJB with a **@schedule** that isn't injected anywhere.

I checked the Derby database documentation. There was no unused space to reclaim. I found out that the thousands 1MB files in the log directory are the Write Ahead Log. Opening the ejbtimer database from the Derby command line tool took as long as Glassfish to start. This single application generates arround 50MB of WAL per day. It seems a lot. For an unknown reason, the WAL files are not deleted after a checkpoint or a shutdown (tested with Derby 10.9 and 10.13). Moreover, Derby seems to parse the 10GB WAL like it does in recovery mode despite that the data is written and up to date on the disk.

So I simply shutdown the database, deleted the WAL manually, ensured that the data was still there and Glassfish now start within a couple of seconds.

References : 
https://db.apache.org/derby/papers/DerbyTut/embedded_intro.html#shutdown
https://db.apache.org/derby/docs/10.9/ref/rrefattrib16471.html
